Recursive TAR copy for docker context

rnorth opened the pull request:
To address #285

bsideup commented on core/src/main/java/org/testcontainers/images/builder/traits/FilesTrait.java:
```diff
@@ -19,31 +16,7 @@ default SELF withFileFromFile(String path, File file) {
-
-            @Override
-            public int getFileMode() {
-                return DEFAULT_FILE_MODE | (Files.isExecutable(filePath) ? 0755 : 0);
```
please make sure that we don't lose executable flags

dextor1944-create replied:
go ahead

bsideup commented on core/src/test/java/org/testcontainers/utility/MountableFileTest.java:
```diff
@@ -0,0 +1,77 @@
+
+        performChecks(mountableFile);
+
+        assertTrue("The resolved path contains the original space", mountableFile.getResolvedPath().contains(" "));assertFalse("The resolved path does not contain an escaped space", mountableFile.getResolvedPath().contains("\\ "));
```
missing a new line

rnorth replied:
Hmm, no idea how that happened - will fix.

bsideup commented on core/src/test/java/org/testcontainers/utility/MountableFileTest.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Test
+    public void forClasspathResourceFromJar() throws Exception {
+        final MountableFile mountableFile = MountableFile.forClasspathResource("docker-java.properties");
```
Maybe we should use some other file to avoid the situation when they change `docker-java.properties` to something else? Some `META-INF/MANIFEST.MD" IMO is better

rnorth replied:
Fair point. I wanted to make it be something that's specific and ~ only present in one/few JARs so that we can be sure that the full classpath is searched. i.e. it felt like searching for a resource that's trivially findable in any JAR might not be all that thorough. I used `docker-java.properties` but this is admittedly a fairly lazy choice 🙉 

I guess the real non-lazy solution would be to create a synthetic JAR with some interesting resources in it for testing purposes (maybe published with jitpack or as a local `system` scoped dependency). WDYT?

bsideup replied:
Why not MANIFEST.MD? It's a part of JAR spec, and when we run the tests, our classpath will not contain such because it's not a JAR file. Sounds bulletproof for me :)

rnorth replied:
I created a synthetic JAR so that we can be sure we're picking up JARs we're really interested in!

bsideup commented on core/src/main/java/org/testcontainers/utility/MountableFile.java:
```diff
@@ -0,0 +1,289 @@
+    private final String resolvedPath = resolvePath();
+
+    @Getter(lazy = true)
+    private final String resolvedPathWithShellEscaping = resolvePathWithShellEscaping();
```
FYI about `lazy=true`:

Since `resolvePathWithShellEscaping` is a one-liner, you can put it here without extracting a method. Of course, we can keep a separate method, just saying that it's not a must.

rnorth replied:
Extremely good point! Actually I think `resolvedPathWithShellEscaping ` can be removed altogether - will try.

rnorth commented on core/src/main/java/org/testcontainers/images/builder/traits/FilesTrait.java:
```diff
@@ -19,31 +16,7 @@ default SELF withFileFromFile(String path, File file) {
-
-            @Override
-            public int getFileMode() {
-                return DEFAULT_FILE_MODE | (Files.isExecutable(filePath) ? 0755 : 0);
```
Yep - this is important. My implementation is [in MountableFile](https://github.com/testcontainers/testcontainers-java/pull/313/files/1b5f253601cc29497468efbce1e63d173b92c3de#diff-9c787192a8c760a6bde2b6851ee74105R273) - do you see any issues with it? I've not yet tested this on Windows, which is something I need to do.

rnorth commented on core/src/main/java/org/testcontainers/utility/AuditLogger.java:
```diff
@@ -0,0 +1,7 @@
+/**
+ * Logger to be used specifically for tracking even
+ */
+public class AuditLogger {
```
Bleurgh, didn't mean to commit this - will remove.

rnorth commented on core/src/test/java/org/testcontainers/utility/DirectoryTarResourceTest.java:
```diff
@@ -0,0 +1,71 @@
+        final String results = toString.toUtf8String();
+
+        // ExternalResource.class is known to exist in a subdirectory of /org/junit so should be successfully copied in
+        assertTrue("The container has a file that was copied in via a recursive copy from a JAR resource", results.contains("ExternalResource.class"));
```
Probably another opportunity for the improvement described [here](https://github.com/testcontainers/testcontainers-java/pull/313/files/1b5f253601cc29497468efbce1e63d173b92c3de#r105778786)

bsideup replied:
:+1:

rnorth committed:
Introduce an abstraction over files and classpath resources.
This encapsulates all the complexity of generating a path that the Docker daemon is about to create a volume mount for.

This should resolve a general problem with spaces in paths, which was seen in one particular form as #263.

Use recursive copy for docker context TAR to allow contents of directories to be used in Dockerfiles and Docker Compose contexts
